src/Microsoft.AspNetCore.Routing/Dispatcher/RouteTemplateUrlGenerator.cs

```diff
@@ -37,14 +37,13 @@ public string GenerateUrl(HttpContext httpContext, object values)
                 throw new ArgumentNullException(nameof(values));
             }
 
-            var address = _addressSelector.SelectAddress(new DispatcherValueCollection(values));
+            var address = _addressSelector.SelectAddress(new DispatcherValueCollection(values)) as ITemplateAddress;
```

Comment:
Contract of the `TemplateAddressSelector` returns an `ITemplateAddress + Address`. We don't have sum types tho. Sad panda